tests/Configuration/RawConfiguration/RawConfigurationTest.php

```diff
@@ -0,0 +1,179 @@
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+declare(strict_types=1);
+
+namespace Infection\Tests\Configuration\RawConfiguration;
+
+use Exception;
+use Generator;
+use Infection\Configuration\RawConfiguration\InvalidFile;
+use Infection\Configuration\RawConfiguration\RawConfiguration;
+use PHPUnit\Framework\TestCase;
+use ReflectionClass;
+use Seld\JsonLint\ParsingException;
+
+final class RawConfigurationTest extends TestCase
+{
+    private const FIXTURES_DIR = __DIR__ . '/../../Fixtures/Configuration';
+
+    public function test_it_can_be_instantiated(): void
+    {
+        $path = '/nowhere';
+
+        $config = new RawConfiguration($path);
+
+        $this->assertSame($path, $config->getPath());
+    }
+
+    public function test_its_contents_is_retrieved_lazily(): void
+    {
+        $invalidPath = '/nowhere';
+
+        $config = new RawConfiguration($invalidPath);
+
+        try {
+            $config->getDecodedContents();
+
+            $this->fail('Expected the content to be invalid.');
+        } catch (Exception $exception) {
+            $this->addToAssertionCount(1);
+        }
+
+        $validPath = self::FIXTURES_DIR . '/file.json';
+        $expectedArrayContents = ['foo' => 'bar'];
+
+        $config = new RawConfiguration($validPath);
+        $actualContents = $config->getDecodedContents();
+
+        $this->assertSame($expectedArrayContents, (array) $actualContents);
+    }
+
+    public function test_its_contents_is_retrieved_only_once(): void
+    {
+        $config = new RawConfiguration(self::FIXTURES_DIR . '/file.json');
+        $expectedValue = (object) ['a' => 'b'];
+
+        // Fetch the contents once
+        $config->getDecodedContents();
+
+        $decodedContentsReflection = (new ReflectionClass(RawConfiguration::class))->getProperty('decodedContents');
+        $decodedContentsReflection->setAccessible(true);
+        $decodedContentsReflection->setValue($config, $expectedValue);
+
+        $this->assertSame($expectedValue, $config->getDecodedContents());
+    }
+
+    /**
+     * @dataProvider invalidConfigContentsProvider
+     */
+    public function test_it_cannot_retrieve_or_decode_invalid_contents(
+        string $path,
+        Exception $expectedException
+    ): void {
+        $config = new RawConfiguration($path);
+
+        try {
+            $config->getDecodedContents();
+
+            $this->fail('Expected the config contents to be invalid.');
+        } catch (Exception $exception) {
+            // Continue
+        }
+
+        $this->assertSame(
+            $expectedException->getMessage(),
+            $exception->getMessage()
+        );
+        $this->assertSame(
+            $expectedException->getCode(),
+            $exception->getCode()
+        );
+
+        if (null === $expectedException->getPrevious()) {
```

Comment:
I'm really not a fan of `if `statemetns inside of tests, to assert differnt things. I think its better to split up the tests in to multiple tests.

Reply:
I tend to agree when it's about a case where it's valid in one block and an exception in another. However here this if block is to do extra assertions on the exception thrown. I think splitting it in two test methods will increase make it a bit more confusing